Improved documentation about virtual threads. (#8900)
eclipse/jetty.project: documentation/jetty-documentation/src/main/asciidoc/programming-guide/arch-threads.adoc

//
// ========================================================================
// Copyright (c) 1995-2022 Mort Bay Consulting Pty Ltd and others.
//
// This program and the accompanying materials are made available under the
// terms of the Eclipse Public License v. 2.0 which is available at
// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
// which is available at https://www.apache.org/licenses/LICENSE-2.0.
//
// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
// ========================================================================
//

[[pg-arch-threads]]
=== Jetty Threading Architecture

Writing a performant client or server is difficult, because it should:

* Scale well with the number of processors.
* Be efficient at using processor caches to avoid link:https://en.wikipedia.org/wiki/Parallel_slowdown[parallel slowdown].
* Support multiple network protocols that may have very different requirements; for example, multiplexed protocols such as HTTP/2 introduce new challenges that are not present in non-multiplexed protocols such as HTTP/1.1.

> * Support different application threading models; for example, if a Jetty server invokes server-side application code that is allowed to call blocking APIs, then the Jetty server should not be affected by how long the blocking API call is, and should be able to process other connections or other requests in a timely fashion.

Reviewer: ```suggestion
Support different application threading models; for example, if a Jetty server invokes server-side application code that is allowed to call blocking APIs, then the Jetty server should not be affected by how long the blocking API call takes, and should be able to process other connections or other requests in a timely fashion.
```

Author: Done.

For example, Jetty produces (among others) these tasks:

* A task that wraps a NIO selection event, see the xref:pg-arch-io[Jetty I/O architecture].
* A task that wraps the invocation of application code that may block (for example, the invocation of a Servlet to handle an HTTP request).

A task is typically a `Runnable` object that may implement `org.eclipse.jetty.util.thread.Invocable` to indicate the behavior of the task (in particular, whether the task may block or not).

Once a task has been produced, it may be consumed using these strategies:

* `Produce-Consume`
* `Produce-Execute-Consume`
* `Execute-Produce-Consume`

[[pg-arch-threads-execution-strategy-pc]]
===== Produce-Consume
In the `Produce-Consume` mode, the producer thread loops to produce a task that is run directly by the `Producer Thread`.

[plantuml]
----
skinparam backgroundColor transparent

compact concise "Producer Thread" as PT
hide time-axis

@PT
0 is T1 #lightgreen
1 is "Run T1" #dodgerblue
5 is T2 #lightgreen
6 is "Run T2" #dodgerblue
8 is T3 #lightgreen
9 is "Run T3" #dodgerblue
12 is {hidden}
----

> If the task is a NIO selection event, then this model is the thread-per-selector model which is very CPU core cache efficient, but suffers from the link:http://en.wikipedia.org/wiki/Head-of-line_blocking[head-of-line blocking]: if one of the tasks blocks or runs slowly, then subsequent tasks cannot be produced (and therefore cannot be consumed either) and will pay in latency the cost of running previous, possibly unrelated, tasks.

Reviewer 2: Maybe we should include hints about the Adaptive Strategy:
```suggestion
If the task is a NIO selection event, then this model is the thread-per-selector model which is very CPU core cache efficient, but suffers from the link:http://en.wikipedia.org/wiki/Head-of-line_blocking[head-of-line blocking]: if one of the tasks blocks or runs slowly, then subsequent tasks cannot be produced (and therefore cannot be consumed either) and will pay in latency the cost of running previous, possibly unrelated, tasks.

This strategy should only be used if the produced task is known not to block or the deployment is able to suffer head of line blocking.
```

Author: Slightly modified, but done.

[[pg-arch-threads-execution-strategy-pec]]
===== Produce-Execute-Consume
In the `Produce-Execute-Consume` mode, the `Producer Thread` loops to produce tasks that are submitted to a `java.util.concurrent.Executor` to be run by ``Worker Thread``s different from the `Producer Thread`.

[plantuml]
----
skinparam backgroundColor transparent

compact concise "Producer Thread" as PT
compact concise "Worker Thread 1" as WT1
compact concise "Worker Thread 2" as WT2
compact concise "Worker Thread 3" as WT3
hide time-axis

@PT
0 is T1 #lightgreen
1 is T2 #lightgreen
2 is T3 #lightgreen
3 is T4 #lightgreen
4 is {hidden}

@WT1
1 is "Run T1" #dodgerblue
5 is {hidden}

@WT2
2 is "Run T2" #dodgerblue
4 is "Run T4" #dodgerblue
8 is {hidden}

@WT3
3 is "Run T3" #dodgerblue
6 is {hidden}
----

> The `Executor` implementation typically adds the task to a queue, and dequeues the task when there is a worker thread available to run it.

Reviewer 2: ```suggestion
The `Executor` implementation typically adds the task to a queue, and dequeues the task when there is a worker thread available to run it. Configurations that support virtual threading are available that avoid such task queues.
```

Author: I don't think that is true. A virtual thread is "queued" as a continuation into the scheduler, typically the `ForkJoinPool`, so the generic sentence works for virtual thread too.

Reviewer 2: But... But... Loom is magic! It is not bound by such concerns!

But seriously, i think a mention here that virtual threads are an alternative to thread pooling would be good

Reviewer 3: Loom is trying to make it so that you don't care about those nuances and details.

That's what it seems like (at least) when you read their own documentation and blog posts about it.

Author: This paragraph is not the right section to introduce virtual threads.
I'm just explaining PEC, and how it works.
It is when PEC is used in the context of AdaptiveExecutionStrategy that virtual threads can be used, so it is explained there.

Reviewer 2: Then leave out the typically explanation of what an executor does.

This mode solves the head-of-line blocking discussed in the xref:pg-arch-threads-execution-strategy-pc[`Produce-Consume` section], but suffers of other issue:

> * It is not CPU core cache efficient, as the data available to the producer thread will need to be accessed by another thread that likely won't have that data in its caches.

Reviewer: ```suggestion
* It is not CPU core cache efficient, as the data available to the producer thread will need to be accessed by another thread that likely is going to run on a CPU core that won't have that data in its cache.
```

Author: Done.

By the time thread `Thread 3` takes over task production from `Thread 2`, all the work might already be done.

This mode may also operate similarly to xref:pg-arch-threads-execution-strategy-pec[`Produce-Execute-Consume`] when the take over production task always finds a free CPU core immediately (for example, in a mostly idle server): thread `Thread 1` will produce a task, yield production to `Thread 2` while `Thread 1` is running the task; `Thread 2` will produce a task, yield production to `Thread 3` while `Thread 2` is running the task, etc.

Differently from `Produce-Execute-Consume`, here production happens on different threads, but the advantage is that the task is run by the same thread that produced it (which is CPU core cache efficient).

[[pg-arch-threads-execution-strategy-adaptive]]
===== Adaptive Execution Strategy
The modes of task consumption discussed above are captured by the `org.eclipse.jetty.util.thread.ExecutionStrategy` interface, with an additional implementation that also takes into account the behavior of the task when the task implements `Invocable`.

> For example, a task that declares itself as non-blocking could be consumed using the `Produce-Consume` mode, since there is no risk to stop production because the task will not block.

Reviewer 2: ```suggestion
For example, a task that declares itself as non-blocking can be consumed using the `Produce-Consume` mode, since there is no risk to stop production because the task will not block.
```

Author: Done.

Conversely, a task that declares itself as blocking will stop production, and therefore must be consumed using either the `Produce-Execute-Consume` mode or the `Execute-Produce-Consume` mode.
Deciding between these two modes depends on whether there is a free thread immediately available to take over production, and this is captured by the `org.eclipse.jetty.util.thread.TryExecutor` interface.

> An implementation of `TryExecutor` could be asked whether it can immediately allocate a free thread to run a task, as opposed to a normal `Executor`, that can only queue the task in the hope that there will be a thread in the near future that could run the task.

Reviewer 2: ```suggestion
An implementation of `TryExecutor` can be asked whether a thread can be immediately allocated to run a task, as opposed to a normal `Executor`, that can only queue the task in the expectation that there will be a thread available in the near future to run the task.
```

Author: Done.

> 

Reviewer 2: ```suggestion

If a task declares that it can run in either blocking or non blocking mode, then the strategy may choose to use Produce-Consume with the additional invocation flag to force non-blocking mode; or it may choose one of the other strategies used for blocking tasks.
```

Author: I left this out on purpose to avoid to go too deep into the implementation details.
For example, I do not mention the EITHER invocation type, the PRODUCE_INVOKE_CONSUME mode, etc.
For now I'd leave it out.

Reviewer 2: I think it is important to add at least a small mention like this. Else it appears that the documentation does not describe the actual implementation. It doesn't need a deep dive, just a mention.

Author: Your suggestion is just too obscure for those that don't know the implementation.
99.999% of usages will be either blocking or non-blocking, "either" is really for super-special cases and if a developer is there, reading the code/javadocs is better than reading the documentation.

Reviewer 2: I still think that EITHER and PIC needs to be mentioned somewhere in the doco, otherwise it looks like it is for a different version of the code.

`QueuedThreadPool` can be configured with a `maxThreads` value.

However, some of the Jetty components (such as the xref:pg-arch-io-selector-manager[selectors]) permanently steal threads for their internal use, or rather `QueuedThreadPool` leases some threads to these components.
These threads are reported by `QueuedThreadPool.leasedThreads` and are not available to run application code.

`QueuedThreadPool` can be configured with a `reservedThreads` value.
This value represents the maximum number of threads that can be reserved and used by the `TryExecutor` implementation.

[[pg-arch-threads-thread-pool-virtual-threads]]
===== Virtual Threads
Virtual threads have been introduced in Java 19 as a preview feature.

NOTE: In Java versions where virtual threads are a preview feature, remember to add `+--enable-preview+` to the command line options to use virtual threads.

`QueuedThreadPool` can be configured to use virtual threads via:

[source,java,indent=0]
----
include::{doc_code}/org/eclipse/jetty/docs/programming/QueuedThreadPoolDocs.java[tags=virtual-threads]

> ----

Reviewer 2: ```suggestion
----
[NOTE]
====
If virtual thread are enabled in `QueuedThreadPool`, then the heuristic for determining the number of reserved threads will return 0, so that a virtual thread will always be used for blocking tasks.   If a deployment wishes to use enable the use of Execute-Produce-Consume strategy for blocking tasks, then a specific number of reserved threads must be configured. 
====
```

Author: Modified and expanded, but done.

===== Virtual Threads
Virtual threads have been introduced in Java 19 as a preview feature.

NOTE: In Java versions where virtual threads are a preview feature, remember to add `+--enable-preview+` to the command line options to use virtual threads.

`QueuedThreadPool` can be configured to use virtual threads via:

[source,java,indent=0]
----
include::{doc_code}/org/eclipse/jetty/docs/programming/QueuedThreadPoolDocs.java[tags=virtual-threads]
----

`AdaptiveExecutionStrategy` makes use of this setting when it determines that a task should be run with the xref:pg-arch-threads-execution-strategy-pec[`Produce-Execute-Consume` mode]: rather than submitting the task to `QueuedThreadPool` to be run in a platform thread, it runs the task in a new virtual thread.

[NOTE]
====
Enabling virtual threads in `QueuedThreadPool` will default the number of reserved threads to zero, unless the number of reserved threads is explicitly configured to a positive value.

Defaulting the number of reserved threads to zero ensures that the xref:pg-arch-threads-execution-strategy-pec[Produce-Execute-Consume mode] is always used, which means that virtual threads will always be used for blocking tasks.

> ====

Reviewer 3: A common question that comes up is how threading is used during an HTTP exchange, is this the document/place to discuss this?

Author: Not here particularly, as this is a generic section that works for both client and server (so maybe 2 sections, one for the client, one for the server), but yes it would be nice to have.
Not in this PR though.
